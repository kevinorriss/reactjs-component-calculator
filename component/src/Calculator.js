import React from 'react'
import mexp from 'math-expression-evaluator'
import ControlKey from './components/ControlKey'
import NumberKey from './components/NumberKey'
import OperatorKey from './components/OperatorKey'
import './styles.css'

class Calculator extends React.Component {
    constructor(props) {
        super(props)
        this.handleNumericKeyEvent = this.handleNumericKeyEvent.bind(this)
        this.handleOperatorKeyEvent = this.handleOperatorKeyEvent.bind(this)
        this.handleClearEvent = this.handleClearEvent.bind(this)
        this.handleClearEntryEvent = this.handleClearEntryEvent.bind(this)
        this.handleDeleteEvent = this.handleDeleteEvent.bind(this)
        this.handleDecimalEvent = this.handleDecimalEvent.bind(this)
        this.handleEqualsEvent = this.handleEqualsEvent.bind(this)

        this.state = {
            left: null,
            operator: null,
            right: null,
            answer: null
        }
    }

    handleNumericKeyEvent(number) {
        let value = this.getCurrentValue()
        this.setCurrentValue((value === null || value === '0' || isNaN(value) || !isFinite(value)) ? number : value + number)
    }

    handleOperatorKeyEvent(operator) {
        this.setState((prevState) => ({
            left: prevState.left || prevState.answer,
            operator,
            right: prevState.right,
            answer: null
        }))
    }

    handleClearEvent() {
        this.setState({
            left: null,
            operator: null,
            right: null,
            answer: null
        })
    }

    handleClearEntryEvent() {
        this.setState((prevState) => ({
            left: prevState.operator ? prevState.left : null,
            operator: prevState.operator,
            right: null,
            answer: null
        }))
    }

    handleDeleteEvent() {
        const value = this.getCurrentValue()
        this.setCurrentValue(value === null || RegExp('^-?.$').test(value) ? null : value.substring(0, value.length - 1))
    }

    handleDecimalEvent() {
        const value = this.getCurrentValue()
        if (value === null || !value.match(/\./)) {
            this.setCurrentValue((value === null || value === '0' || isNaN(value) || !isFinite(value)) ? '0.' : value + '.')
        }
    }

    handleEqualsEvent() {
        if (this.state.right !== null) {
            this.setState({
                left: null,
                operator: null,
                right: null,
                answer: mexp.eval(this.getDisplayText().replace('%', 'Mod')).toString()
            })
        }
    }

    getCurrentValue() {
        return this.state.operator ? this.state.right : this.state.left || this.state.answer
    }

    setCurrentValue(value) {
        this.setState((prevState) => ({
            left: prevState.operator ? prevState.left : value,
            operator: prevState.operator,
            right: prevState.operator ? value : prevState.right,
            answer: null
        }))
    }

    getDisplayText() {
        if (this.state.answer) {
            return this.state.answer
        }
        return (`${this.state.left || '0'} ${this.state.operator || ''} ${this.state.right || ''}`).trim()
    }

    render() {
        return (
            <div className="calculator">
                <div className="display">
                    <div className="display--inner">
                        <h4 className="display--value">{this.getDisplayText()}</h4>
                    </div>
                </div>
                <table className="keypad">
                    <tbody>
                        <tr>
                            <td width="25%"><ControlKey text="C" keys={['Escape']} onClick={this.handleClearEvent}/></td>
                            <td width="50%" colSpan={2}><ControlKey text="CE" keys={['Delete']} onClick={this.handleClearEntryEvent}/></td>
                            <td width="25%"><OperatorKey text="%" onClick={this.handleOperatorKeyEvent}/></td>
                        </tr>
                        <tr>
                            <td><NumberKey number={1} onClick={this.handleNumericKeyEvent}/></td>
                            <td><NumberKey number={2} onClick={this.handleNumericKeyEvent}/></td>
                            <td><NumberKey number={3} onClick={this.handleNumericKeyEvent}/></td>
                            <td><OperatorKey text="/" onClick={this.handleOperatorKeyEvent}/></td>
                        </tr>
                        <tr>
                            <td><NumberKey number={4} onClick={this.handleNumericKeyEvent}/></td>
                            <td><NumberKey number={5} onClick={this.handleNumericKeyEvent}/></td>
                            <td><NumberKey number={6} onClick={this.handleNumericKeyEvent}/></td>
                            <td><OperatorKey text="*" onClick={this.handleOperatorKeyEvent}/></td>
                        </tr>
                        <tr>
                            <td><NumberKey number={7} onClick={this.handleNumericKeyEvent}/></td>
                            <td><NumberKey number={8} onClick={this.handleNumericKeyEvent}/></td>
                            <td><NumberKey number={9} onClick={this.handleNumericKeyEvent}/></td>
                            <td><OperatorKey text="-" onClick={this.handleOperatorKeyEvent}/></td>
                        </tr>
                        <tr>
                            <td><ControlKey text="del" keys={['Backspace']} onClick={this.handleDeleteEvent}/></td>
                            <td><NumberKey number={0} onClick={this.handleNumericKeyEvent}/></td>
                            <td><ControlKey text="." keys={['.']} onClick={this.handleDecimalEvent}/></td>
                            <td><OperatorKey text="+" onClick={this.handleOperatorKeyEvent}/></td>
                        </tr>
                        <tr>
                            <td colSpan={4}><ControlKey text="=" keys={['=', 'Enter']} onClick={this.handleEqualsEvent} className="button--control-equals"/></td>
                        </tr>
                    </tbody>
                </table>
            </div>
        )
    }
}

export default Calculator
