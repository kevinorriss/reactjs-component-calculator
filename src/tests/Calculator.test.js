import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import Calculator from '../../component/src/Calculator'

let wrapper
beforeEach(() => {
    wrapper = mount(<Calculator/>)
})

const getNumberKey = (number) => {
    return wrapper.find('.button--number').filterWhere(c => c.text() == number).first()
}

const getControlKey = (text) => {
    return wrapper.find('.button--control').filterWhere(c => c.text() === text).first()
}

const getOperatorKey = (text) => {
    return wrapper.find('.button--operator').filterWhere(c => c.text() === text).first()
}

const getDisplayText = () => {
    return wrapper.find('.display--value').text()
}

it('should render Calculator correctly', () => {
    const component = renderer.create(<Calculator/>).toJSON()
    expect(component).toMatchSnapshot()
})

it('should append digits to current value', () => {
    getNumberKey(1).simulate('click')
    getNumberKey(3).simulate('click')
    getNumberKey(5).simulate('click')

    expect(getDisplayText()).toEqual('135')
})

it('should accept decimal values', () => {
    getNumberKey(4).simulate('click')
    getControlKey('.').simulate('click')
    getNumberKey(5).simulate('click')

    expect(getDisplayText()).toEqual('4.5')
})

it('should reject multiple decimal characters', () => {
    const decimal = getControlKey('.')
    const key4 = getNumberKey(4)
    const key5 = getNumberKey(5)
    const key6 = getNumberKey(6)

    key4.simulate('click')
    key5.simulate('click')
    decimal.simulate('click')
    decimal.simulate('click')
    key6.simulate('click')
    expect(getDisplayText()).toEqual('45.6')
})

it('should reject 2nd decimal characters after numbers', () => {
    const decimal = getControlKey('.')
    const key4 = getNumberKey(4)
    const key5 = getNumberKey(5)
    const key6 = getNumberKey(6)

    key4.simulate('click')
    decimal.simulate('click')
    key5.simulate('click')
    decimal.simulate('click')
    key6.simulate('click')
    expect(getDisplayText()).toEqual('4.56')
})

it ('should reject multiple 0 keys', () => {
    const zero = getNumberKey(0)
    zero.simulate('click')
    zero.simulate('click')
    
    expect(getDisplayText()).toEqual('0')
})

it ('should accept multiple 0 keys after non-zero key', () => {
    const zero = getNumberKey(0)
    getNumberKey(1).simulate('click')
    zero.simulate('click')
    zero.simulate('click')
    
    expect(getDisplayText()).toEqual('100')
})

it('should clear calculation on clear button press', () => {
    getNumberKey(7).simulate('click')
    getNumberKey(8).simulate('click')
    getControlKey('.').simulate('click')
    getNumberKey(9).simulate('click')

    expect(getDisplayText()).toEqual('78.9')

    getControlKey('C').simulate('click')

    expect(getDisplayText()).toEqual('0')
})

it('should clear left value on clear entry button press', () => {
    getNumberKey(7).simulate('click')
    expect(getDisplayText()).toEqual('7')

    getControlKey('CE').simulate('click')
    expect(getDisplayText()).toEqual('0')
})

it('should delete right-most character on delete key press', () => {
    getNumberKey(1).simulate('click')
    getNumberKey(2).simulate('click')
    expect(getDisplayText()).toEqual('12')

    getControlKey('del').simulate('click')
    expect(getDisplayText()).toEqual('1')
})

it('should delete right-most character on delete key press (decimal)', () => {
    getNumberKey(1).simulate('click')
    getNumberKey(2).simulate('click')
    getControlKey('.').simulate('click')
    expect(getDisplayText()).toEqual('12.')

    getControlKey('del').simulate('click')
    expect(getDisplayText()).toEqual('12')
})

it('should reset to 0 when delete key pressed on single digit value', () => {
    getNumberKey(2).simulate('click')
    expect(getDisplayText()).toEqual('2')

    getControlKey('del').simulate('click')
    expect(getDisplayText()).toEqual('0')
})

it('should do nothing when equals pressed before complete calculation', () => {
    getNumberKey(2).simulate('click')
    getControlKey('=').simulate('click')
    expect(getDisplayText()).toEqual('2')
})

it('should handle addtion calculation', () => {
    getNumberKey(1).simulate('click')
    getControlKey('.').simulate('click')
    getNumberKey(2).simulate('click')
    getOperatorKey('+').simulate('click')
    getNumberKey(3).simulate('click')
    getControlKey('.').simulate('click')
    getNumberKey(4).simulate('click')
    expect(getDisplayText()).toEqual('1.2 + 3.4')

    getControlKey('=').simulate('click')
    expect(getDisplayText()).toEqual('4.6')
})

it('should handle subtraction calculation', () => {
    getNumberKey(1).simulate('click')
    getControlKey('.').simulate('click')
    getNumberKey(2).simulate('click')
    getOperatorKey('-').simulate('click')
    getNumberKey(3).simulate('click')
    getControlKey('.').simulate('click')
    getNumberKey(4).simulate('click')
    expect(getDisplayText()).toEqual('1.2 - 3.4')

    getControlKey('=').simulate('click')
    expect(getDisplayText()).toEqual('-2.2')
})

it('should handle multiplication calculation', () => {
    getNumberKey(1).simulate('click')
    getControlKey('.').simulate('click')
    getNumberKey(2).simulate('click')
    getOperatorKey('*').simulate('click')
    getNumberKey(3).simulate('click')
    getControlKey('.').simulate('click')
    getNumberKey(4).simulate('click')
    expect(getDisplayText()).toEqual('1.2 * 3.4')

    getControlKey('=').simulate('click')
    expect(getDisplayText()).toEqual('4.08')
})

it('should handle division calculation', () => {
    getNumberKey(1).simulate('click')
    getNumberKey(0).simulate('click')
    getOperatorKey('/').simulate('click')
    getNumberKey(4).simulate('click')
    expect(getDisplayText()).toEqual('10 / 4')

    getControlKey('=').simulate('click')
    expect(getDisplayText()).toEqual('2.5')
})

it('should handle modulus calculation', () => {
    getNumberKey(1).simulate('click')
    getNumberKey(0).simulate('click')
    getOperatorKey('%').simulate('click')
    getNumberKey(3).simulate('click')
    expect(getDisplayText()).toEqual('10 % 3')

    getControlKey('=').simulate('click')
    expect(getDisplayText()).toEqual('1')
})

it('should reset to 0 when delete key pressed on single digit negative value', () => {
    getOperatorKey('-').simulate('click')
    getNumberKey(5).simulate('click')
    getControlKey('=').simulate('click')

    expect(getDisplayText()).toEqual('-5')

    getControlKey('del').simulate('click')
    expect(getDisplayText()).toEqual('0')
})

it('should handle operator changing', () => {
    getNumberKey(5).simulate('click')
    getOperatorKey('-').simulate('click')
    expect(getDisplayText()).toEqual('5 -')

    getOperatorKey('+').simulate('click')
    expect(getDisplayText()).toEqual('5 +')

    getNumberKey(6).simulate('click')
    expect(getDisplayText()).toEqual('5 + 6')

    getOperatorKey('*').simulate('click')
    expect(getDisplayText()).toEqual('5 * 6')
})

it('should use answer as new calculation left value', () => {
    getNumberKey(2).simulate('click')
    getOperatorKey('+').simulate('click')
    getNumberKey(3).simulate('click')
    getControlKey('=').simulate('click')
    getOperatorKey('*').simulate('click')
    expect(getDisplayText()).toEqual('5 *')
})

it('should do nothing when pressing delete after operator (no right value)', () => {
    getNumberKey(2).simulate('click')
    getOperatorKey('+').simulate('click')
    getControlKey('del').simulate('click')
    expect(getDisplayText()).toEqual('2 +')
})