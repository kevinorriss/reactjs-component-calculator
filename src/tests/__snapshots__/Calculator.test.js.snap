// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Calculator correctly 1`] = `
<div
  className="calculator"
>
  <div
    className="display"
  >
    <div
      className="display--inner"
    >
      <h4
        className="display--value"
      >
        0
      </h4>
    </div>
  </div>
  <table
    className="keypad"
  >
    <tbody>
      <tr>
        <td
          width="25%"
        >
          <div
            className="button button--control"
            onClick={[Function]}
          >
            <div>
              C
            </div>
          </div>
        </td>
        <td
          colSpan={2}
          width="50%"
        >
          <div
            className="button button--control"
            onClick={[Function]}
          >
            <div>
              CE
            </div>
          </div>
        </td>
        <td
          width="25%"
        >
          <div
            className="button button--operator"
            onClick={[Function]}
          >
            <div>
              %
            </div>
          </div>
        </td>
      </tr>
      <tr>
        <td>
          <div
            className="button button--number"
            onClick={[Function]}
          >
            <div>
              1
            </div>
          </div>
        </td>
        <td>
          <div
            className="button button--number"
            onClick={[Function]}
          >
            <div>
              2
            </div>
          </div>
        </td>
        <td>
          <div
            className="button button--number"
            onClick={[Function]}
          >
            <div>
              3
            </div>
          </div>
        </td>
        <td>
          <div
            className="button button--operator"
            onClick={[Function]}
          >
            <div>
              /
            </div>
          </div>
        </td>
      </tr>
      <tr>
        <td>
          <div
            className="button button--number"
            onClick={[Function]}
          >
            <div>
              4
            </div>
          </div>
        </td>
        <td>
          <div
            className="button button--number"
            onClick={[Function]}
          >
            <div>
              5
            </div>
          </div>
        </td>
        <td>
          <div
            className="button button--number"
            onClick={[Function]}
          >
            <div>
              6
            </div>
          </div>
        </td>
        <td>
          <div
            className="button button--operator"
            onClick={[Function]}
          >
            <div>
              *
            </div>
          </div>
        </td>
      </tr>
      <tr>
        <td>
          <div
            className="button button--number"
            onClick={[Function]}
          >
            <div>
              7
            </div>
          </div>
        </td>
        <td>
          <div
            className="button button--number"
            onClick={[Function]}
          >
            <div>
              8
            </div>
          </div>
        </td>
        <td>
          <div
            className="button button--number"
            onClick={[Function]}
          >
            <div>
              9
            </div>
          </div>
        </td>
        <td>
          <div
            className="button button--operator"
            onClick={[Function]}
          >
            <div>
              -
            </div>
          </div>
        </td>
      </tr>
      <tr>
        <td>
          <div
            className="button button--control"
            onClick={[Function]}
          >
            <div>
              del
            </div>
          </div>
        </td>
        <td>
          <div
            className="button button--number"
            onClick={[Function]}
          >
            <div>
              0
            </div>
          </div>
        </td>
        <td>
          <div
            className="button button--control"
            onClick={[Function]}
          >
            <div>
              .
            </div>
          </div>
        </td>
        <td>
          <div
            className="button button--operator"
            onClick={[Function]}
          >
            <div>
              +
            </div>
          </div>
        </td>
      </tr>
      <tr>
        <td
          colSpan={4}
        >
          <div
            className="button button--control button--control-equals"
            onClick={[Function]}
          >
            <div>
              =
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
